### ISSUE TITLE
[Omniscia audit] Inexistent Minimum Purchase Return

### ISSUE THREAD

**Issue by hiroshi-yamamoto-dublr** [dublr/dublr#5]
## DUB-04M: Inexistent Minimum Purchase Return

| Type | Severity | Location |
| ---- | -------- | -------- |
| [Logical Fault](/appendix/finding-types#logical-fault) | <img class="o-severity o-major" src="https://omniscia.io/report-assets/major.png" /> | [Dublr.sol:L389](https://github.com/dublr/dublr/blob/b84ea8edd8c1ab6a3f1abbc34467fa2ce4b21526/contracts/main/Dublr/Dublr.sol#L389) |

### Description:

The `buy` function does not contain a minimum purchase return argument acting as "slippage" rendering it extremely prone to race-conditions, especially in a first-come first-serve DEX environment. As an example, if the order book is stale and a single party contains the majority of low-price orders, they can swap the price of the bottom-price orders once they detect a large buy transaction thus causing the buyer to pay an exorbitant amount of funds for a small amount of Dublr.

### Impact:

In a relatively stale order book scenario, the lowest-price market maker is in complete control of the price the buyer will ultimately pay per Dublr unit thus compromising the DEX's integrity.

### Example:

```solidity
function buy(bool allowMinting) public payable override(IDublrDEX) stateUpdater {
```

### Recommendation:

We advise the contract to introduce a slippage mechanism guaranteeing that the buyer has received at least an X amount of Dublr tokens in return for their payment of `msg.value` thus eliminating race conditions such as the one described above. For more information on the topic, consult the Uniswap V2 and V3 documentation as well as the rationale behind their slippage arguments.

**Comment by hiroshi-yamamoto-dublr** [comment]
Fixed in b85b3b5.

**Comment by hiroshi-yamamoto-dublr** [comment]
Improved the behavior in 814de86d8cb2bb18cc201f52ee92e58bb7944bbf. If the minimum purchase amount is not met because buying sell orders was terminated prematurely due to out of gas criterion, then the transaction should revert with "Out of gas" rather than "Too much slippage".
